File: iminling-core/src/main/java/com/iminling/core/config/mybatis/MybatisMapperRefresh.java
package com.iminling.core.config.mybatis;

import com.google.common.collect.Lists;
import org.apache.ibatis.binding.MapperRegistry;
import org.apache.ibatis.builder.xml.XMLMapperBuilder;
import org.apache.ibatis.builder.xml.XMLMapperEntityResolver;
import org.apache.ibatis.executor.ErrorContext;
import org.apache.ibatis.executor.keygen.SelectKeyGenerator;
import org.apache.ibatis.io.Resources;
import org.apache.ibatis.mapping.MappedStatement;
import org.apache.ibatis.parsing.XNode;
import org.apache.ibatis.parsing.XPathParser;
import org.apache.ibatis.session.Configuration;
import org.apache.ibatis.session.SqlSessionFactory;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.core.io.FileSystemResource;
import org.springframework.core.io.Resource;
import org.springframework.core.io.UrlResource;
import org.springframework.util.ResourceUtils;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.Field;
import java.util.*;
import java.util.concurrent.TimeUnit;

/**
 * mybatis热加载mapper
 * @author yslao@outlook.com
 * @since 2021/3/11
 */
public class MybatisMapperRefresh {

    private static final Logger logger = LoggerFactory.getLogger(MybatisMapperRefresh.class);
    /**
     * 记录jar包存在的mapper
     */
    private static final Map<String, List<Resource>> jarMapper = new HashMap<>();
    private SqlSessionFactory sqlSessionFactory;
    private Resource[] mapperLocations;
    private Configuration configuration;
    /**
     * 记录文件最后更新时间
     */
    private Map<String, Long> file2UpdateTime = new HashMap<>();

    /**
     * xml文件目录
     */
    private Set<String> fileSet;
    /**
     * 延迟加载时间
     */
    private int delaySeconds = 10;
    /**
     * 刷新间隔时间
     */
    private int sleepSeconds = 20;

    public MybatisMapperRefresh(Resource[] mapperLocations, SqlSessionFactory sqlSessionFactory, int sleepSeconds) {
        this.mapperLocations = mapperLocations.clone();
        this.sqlSessionFactory = sqlSessionFactory;
        this.sleepSeconds = sleepSeconds;
        this.configuration = sqlSessionFactory.getConfiguration();
        this.startRefresh();
    }

    public MybatisMapperRefresh(Resource[] mapperLocations, SqlSessionFactory sqlSessionFactory) {
        this.mapperLocations = mapperLocations.clone();
        this.sqlSessionFactory = sqlSessionFactory;
        this.configuration = sqlSessionFactory.getConfiguration();
        this.startRefresh();
    }

    /**
     * 启动刷新功能
     */
    private void startRefresh() {
        logger.info("start refresh mapper.");
        // final GlobalConfig globalConfig = GlobalConfigUtils.getGlobalConfig(configuration);
        /*
         * 启动 XML 热加载
         */
        long beforeTime = System.currentTimeMillis();
        new Thread(() -> {
            if (fileSet == null) {
                fileSet = new HashSet<>();
                if (mapperLocations != null) {
                    for (Resource mapperLocation : mapperLocations) {
                        try {
                            if (ResourceUtils.isJarURL(mapperLocation.getURL())) {
                                String key = new UrlResource(ResourceUtils.extractJarFileURL(mapperLocation.getURL()))
                                        .getFile().getPath();
                                fileSet.add(key);
                                file2UpdateTime.put(key, beforeTime);
                                if (jarMapper.get(key) != null) {
                                    jarMapper.get(key).add(mapperLocation);
                                } else {
                                    List<Resource> resourcesList = new ArrayList<>();
                                    resourcesList.add(mapperLocation);
                                    jarMapper.put(key, resourcesList);
                                }
                            } else {
                                fileSet.add(mapperLocation.getFile().getPath());
                                file2UpdateTime.put(mapperLocation.getFile().getPath(), beforeTime);
                            }
                        } catch (IOException ioException) {
                            ioException.printStackTrace();
                        }
                    }
                }
            }
            /*try {
                TimeUnit.SECONDS.sleep(delaySeconds);
            } catch (InterruptedException interruptedException) {
                interruptedException.printStackTrace();
            }*/
            do {
                try {
                    for (String filePath : fileSet) {
                        File file = new File(filePath);
                        long lastModified = file2UpdateTime.get(filePath);
                        if (file.isFile() && file.lastModified() > lastModified) {
                            file2UpdateTime.put(filePath, file.lastModified());
                            List<Resource> removeList = jarMapper.get(filePath);
                            if (removeList != null && !removeList.isEmpty()) {
                                for (Resource resource : removeList) {
                                    refresh(resource);
                                }
                            } else {
                                refresh(new FileSystemResource(file));
                            }
                        }
                    }
                } catch (Exception exception) {
                    logger.error("exception", exception);
                }
                try {
                    TimeUnit.SECONDS.sleep(sleepSeconds);
                } catch (InterruptedException interruptedException) {
                    interruptedException.printStackTrace();
                }

            } while (true);
        }, "mybatis refreshMapper").start();
    }

    /**
     * 刷新mapper
     *
     * @throws Exception
     */
    @SuppressWarnings("rawtypes")
    private void refresh(Resource resource) throws ClassNotFoundException, NoSuchFieldException, IllegalAccessException {
        this.configuration = sqlSessionFactory.getConfiguration();
        boolean isSupper = configuration.getClass().getSuperclass() == Configuration.class;
        try {
            Field loadedResourcesField = isSupper ? configuration.getClass().getSuperclass().getDeclaredField("loadedResources")
                    : configuration.getClass().getDeclaredField("loadedResources");
            loadedResourcesField.setAccessible(true);
            Set loadedResourcesSet = ((Set) loadedResourcesField.get(configuration));
            XPathParser xPathParser = new XPathParser(resource.getInputStream(), true, configuration.getVariables(),
                    new XMLMapperEntityResolver());
            XNode context = xPathParser.evalNode("/mapper");
            String namespace = context.getStringAttribute("namespace");
            Field field = MapperRegistry.class.getDeclaredField("knownMappers");
            field.setAccessible(true);
            Map mapConfig = (Map) field.get(configuration.getMapperRegistry());
            // Collection<String> mappedStatementNames = configuration.getMappedStatementNames();

            mapConfig.remove(Resources.classForName(namespace));
            loadedResourcesSet.remove(resource.toString());
            configuration.getCacheNames().remove(namespace);

            cleanParameterMap(context.evalNodes("/mapper/parameterMap"), namespace);
            cleanResultMap(context.evalNodes("/mapper/resultMap"), namespace);
            cleanKeyGenerators(context.evalNodes("insert|update|select"), namespace);
            cleanSqlElement(context.evalNodes("/mapper/sql"), namespace);
            XMLMapperBuilder xmlMapperBuilder = new XMLMapperBuilder(resource.getInputStream(),
                    sqlSessionFactory.getConfiguration(),
                    resource.toString(), sqlSessionFactory.getConfiguration().getSqlFragments());
            xmlMapperBuilder.parse();
            logger.info("refresh: '" + resource + "', success!");
        } catch (IOException e) {
            logger.error("Refresh IOException :" + e.getMessage());
        } finally {
            ErrorContext.instance().reset();
        }
    }

    /**
     * 清理parameterMap
     *
     * @param list
     * @param namespace
     */
    private void cleanParameterMap(List<XNode> list, String namespace) {
        for (XNode parameterMapNode : list) {
            String id = parameterMapNode.getStringAttribute("id");
            configuration.getParameterMaps().remove(namespace + "." + id);
        }
    }

    /**
     * 清理resultMap
     *
     * @param list
     * @param namespace
     */
    private void cleanResultMap(List<XNode> list, String namespace) {
        for (XNode resultMapNode : list) {
            String id = resultMapNode.getStringAttribute("id", resultMapNode.getValueBasedIdentifier());
            configuration.getResultMapNames().remove(id);
            configuration.getResultMapNames().remove(namespace + "." + id);
            clearResultMap(resultMapNode, namespace);
        }
    }

    private void clearResultMap(XNode xNode, String namespace) {
        for (XNode resultChild : xNode.getChildren()) {
            if ("association".equals(resultChild.getName()) || "collection".equals(resultChild.getName())
                    || "case".equals(resultChild.getName())) {
                if (resultChild.getStringAttribute("select") == null) {
                    configuration.getResultMapNames().remove(
                            resultChild.getStringAttribute("id", resultChild.getValueBasedIdentifier()));
                    configuration.getResultMapNames().remove(
                            namespace + "." + resultChild.getStringAttribute("id", resultChild.getValueBasedIdentifier()));
                    if (resultChild.getChildren() != null && !resultChild.getChildren().isEmpty()) {
                        clearResultMap(resultChild, namespace);
                    }
                }
            }
        }
    }

    /**
     * 清理selectKey
     *
     * @param list
     * @param namespace
     */
    private void cleanKeyGenerators(List<XNode> list, String namespace) {
        for (XNode context : list) {
            String id = context.getStringAttribute("id");
            configuration.getKeyGeneratorNames().remove(id + SelectKeyGenerator.SELECT_KEY_SUFFIX);
            configuration.getKeyGeneratorNames().remove(namespace + "." + id + SelectKeyGenerator.SELECT_KEY_SUFFIX);

            Collection<MappedStatement> mappedStatements = configuration.getMappedStatements();
            List<MappedStatement> objects = Lists.newArrayList();
            Iterator<MappedStatement> iterator = mappedStatements.iterator();
            while (iterator.hasNext()) {
                Object next = iterator.next();
                // 跳过mybatis-plus自带的那些查询
                if (next instanceof MappedStatement) {
                    MappedStatement mappedStatement = (MappedStatement) next;
                    if (mappedStatement.getId().equals(namespace + "." + id)) {
                        objects.add(mappedStatement);
                    }
                }
            }
            mappedStatements.removeAll(objects);
        }
    }

    /**
     * 清理sql节点缓存
     *
     * @param list
     * @param namespace
     */
    private void cleanSqlElement(List<XNode> list, String namespace) {
        for (XNode context : list) {
            String id = context.getStringAttribute("id");
            configuration.getSqlFragments().remove(id);
            configuration.getSqlFragments().remove(namespace + "." + id);
        }
    }

}

File: iminling-core/src/main/kotlin/com/iminling/core/config/mybatis/MybatisMapperRefresh.kt
package com.iminling.core.config.mybatis

import com.google.common.collect.Lists
import org.apache.ibatis.binding.MapperRegistry
import org.apache.ibatis.builder.xml.XMLMapperBuilder
import org.apache.ibatis.builder.xml.XMLMapperEntityResolver
import org.apache.ibatis.executor.ErrorContext
import org.apache.ibatis.executor.keygen.SelectKeyGenerator
import org.apache.ibatis.io.Resources
import org.apache.ibatis.mapping.MappedStatement
import org.apache.ibatis.parsing.XNode
import org.apache.ibatis.parsing.XPathParser
import org.apache.ibatis.session.Configuration
import org.apache.ibatis.session.SqlSessionFactory
import org.slf4j.LoggerFactory
import org.springframework.core.io.FileSystemResource
import org.springframework.core.io.Resource
import org.springframework.core.io.UrlResource
import org.springframework.util.ResourceUtils
import java.io.File
import java.util.concurrent.TimeUnit
import kotlin.concurrent.thread

/**
 * mybatis热加载mapper
 * @author  yslao@outlook.com
 * @since  2021/11/29
 */
class MybatisMapperRefresh {

    private val logger = LoggerFactory.getLogger(MybatisMapperRefresh::class.java)

    /**
     * 记录jar包存在的mapper
     */
    private val jarMapper = mutableMapOf<String, List<Resource>>()
    private val sqlSessionFactory: SqlSessionFactory
    private val mapperLocations: Array<Resource>
    private val configuration: Configuration

    /**
     * 记录文件最后更新时间
     */
    private val file2UpdateTime = mutableMapOf<String, Long>()

    /**
     * xml文件目录
     */
    private val fileSet: Set<String> = mutableSetOf()

    /**
     * 延迟加载时间
     */
    private val delaySeconds = 10

    /**
     * 刷新间隔时间
     */
    private var sleepSeconds = 20

    constructor(
        mapperLocations: Array<Resource>,
        sqlSessionFactory: SqlSessionFactory,
        sleepSeconds: Int
    ) {
        this.mapperLocations = mapperLocations.clone()
        this.sqlSessionFactory = sqlSessionFactory
        this.sleepSeconds = sleepSeconds
        configuration = sqlSessionFactory.configuration
        this.startRefresh()
    }

    constructor(mapperLocations: Array<Resource>, sqlSessionFactory: SqlSessionFactory) {
        this.mapperLocations = mapperLocations.clone()
        this.sqlSessionFactory = sqlSessionFactory
        configuration = sqlSessionFactory.configuration
        this.startRefresh()
    }

    /**
     * 启动刷新功能
     */
    private fun startRefresh() {
        logger.info("start refresh mapper.")
        // final GlobalConfig globalConfig = GlobalConfigUtils.getGlobalConfig(configuration);
        /*
         * 启动 XML 热加载
         */
        var beforeTime = System.currentTimeMillis()
        thread(name = "mybatis refreshMapper") {
            if (fileSet.isEmpty()) {
                mapperLocations.forEach {
                    if (ResourceUtils.isJarURL(it.url)) {
                        val key = UrlResource(ResourceUtils.extractJarFileURL(it.url))
                            .file.path
                        fileSet.plus(key)
                        file2UpdateTime[key] = beforeTime
                        if (jarMapper[key] != null) {
                            jarMapper[key]?.plus(it)
                        } else {
                            val resourcesList: MutableList<Resource> = ArrayList()
                            resourcesList.add(it)
                            jarMapper[key] = resourcesList
                        }
                    } else {
                        fileSet.plus(it.file.path)
                        file2UpdateTime[it.file.path] = beforeTime
                    }
                }
            }
            do {
                for (filePath in fileSet) {
                    val file = File(filePath)
                    val lastModified = file2UpdateTime[filePath]!!
                    if (file.isFile && file.lastModified() > lastModified) {
                        file2UpdateTime[filePath] = file.lastModified()
                        val removeList = jarMapper[filePath]
                        if (removeList != null && removeList.isNotEmpty()) {
                            for (resource in removeList) {
                                refresh(resource)
                            }
                        } else {
                            refresh(FileSystemResource(file))
                        }
                    }
                }
                TimeUnit.SECONDS.sleep(sleepSeconds.toLong())
            } while (true)
        }
    }

    private fun refresh(resource: Resource) {
        val isSupper = configuration.javaClass.superclass == Configuration::class.java
        try {
            val loadedResourcesField =
                if (isSupper) configuration.javaClass.superclass.getDeclaredField("loadedResources") else configuration.javaClass.getDeclaredField(
                    "loadedResources"
                )
            loadedResourcesField.isAccessible = true
            val loadedResourcesSet = loadedResourcesField[configuration] as MutableSet<*>
            val xPathParser = XPathParser(
                resource.inputStream, true, configuration.variables,
                XMLMapperEntityResolver()
            )
            val context = xPathParser.evalNode("/mapper")
            val namespace = context.getStringAttribute("namespace")
            val field = MapperRegistry::class.java.getDeclaredField("knownMappers")
            field.isAccessible = true
            val mapConfig = field[configuration.mapperRegistry] as MutableMap<*, *>
            // Collection<String> mappedStatementNames = configuration.getMappedStatementNames();
            mapConfig.remove(Resources.classForName(namespace))
            loadedResourcesSet.remove(resource.toString())
            configuration.cacheNames.remove(namespace)
            cleanParameterMap(context.evalNodes("/mapper/parameterMap"), namespace)
            cleanResultMap(context.evalNodes("/mapper/resultMap"), namespace)
            cleanKeyGenerators(context.evalNodes("insert|update|select"), namespace)
            cleanSqlElement(context.evalNodes("/mapper/sql"), namespace)
            val xmlMapperBuilder = XMLMapperBuilder(
                resource.inputStream,
                sqlSessionFactory.configuration,
                resource.toString(), sqlSessionFactory.configuration.sqlFragments
            )
            xmlMapperBuilder.parse()
            logger.info("refresh: '$resource', success!")
        } catch (e: Exception) {
            logger.error("Refresh Exception :" + e.message)
        } finally {
            ErrorContext.instance().reset()
        }
    }

    /**
     * 清理parameterMap
     *
     * @param list
     * @param namespace
     */
    private fun cleanParameterMap(list: List<XNode>, namespace: String) {
        for (parameterMapNode in list) {
            val id = parameterMapNode.getStringAttribute("id")
            var parameterMap = configuration.getParameterMap("$namespace.$id")
            configuration.parameterMaps.remove(parameterMap)
        }
    }

    /**
     * 清理resultMap
     *
     * @param list
     * @param namespace
     */
    private fun cleanResultMap(list: List<XNode>, namespace: String) {
        for (resultMapNode in list) {
            val id = resultMapNode.getStringAttribute("id", resultMapNode.valueBasedIdentifier)
            configuration.resultMapNames.remove(id)
            configuration.resultMapNames.remove("$namespace.$id")
            clearResultMap(resultMapNode, namespace)
        }
    }

    private fun clearResultMap(xNode: XNode, namespace: String) {
        for (resultChild in xNode.children) {
            if ("association" == resultChild.name || "collection" == resultChild.name || "case" == resultChild.name) {
                if (resultChild.getStringAttribute("select") == null) {
                    configuration.resultMapNames.remove(
                        resultChild.getStringAttribute("id", resultChild.valueBasedIdentifier)
                    )
                    configuration.resultMapNames.remove(
                        namespace + "." + resultChild.getStringAttribute("id", resultChild.valueBasedIdentifier)
                    )
                    if (resultChild.children != null && resultChild.children.isNotEmpty()) {
                        clearResultMap(resultChild, namespace)
                    }
                }
            }
        }
    }

    /**
     * 清理selectKey
     *
     * @param list
     * @param namespace
     */
    private fun cleanKeyGenerators(list: List<XNode>, namespace: String) {
        for (context in list) {
            val id = context.getStringAttribute("id")
            configuration.keyGeneratorNames.remove(id + SelectKeyGenerator.SELECT_KEY_SUFFIX)
            configuration.keyGeneratorNames.remove(namespace + "." + id + SelectKeyGenerator.SELECT_KEY_SUFFIX)
            val mappedStatements = configuration.mappedStatements
            val objects: MutableList<MappedStatement> = Lists.newArrayList()
            val iterator: Iterator<MappedStatement> = mappedStatements.iterator()
            while (iterator.hasNext()) {
                val next: Any = iterator.next()
                // 跳过mybatis-plus自带的那些查询
                if (next is MappedStatement) {
                    if (next.id == "$namespace.$id") {
                        objects.add(next)
                    }
                }
            }
            mappedStatements.removeAll(objects.toSet())
        }
    }

    /**
     * 清理sql节点缓存
     *
     * @param list
     * @param namespace
     */
    private fun cleanSqlElement(list: List<XNode>, namespace: String) {
        for (context in list) {
            val id = context.getStringAttribute("id")
            configuration.sqlFragments.remove(id)
            configuration.sqlFragments.remove("$namespace.$id")
        }
    }
}